{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cee533c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import os\n",
    "import pickle\n",
    "from colections import Counter\n",
    "import pandas as pd\n",
    "import re\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a057d711",
   "metadata": {},
   "outputs": [],
   "source": [
    "#searcHPV workdir\n",
    "workdir = 'MCV_fusion'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7f9b7df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mcvDic = {'VP2': [[465, 1190]],\n",
    " 'VP1': [[1156, 2427]],\n",
    " 'large_T': [[2503, 4722], [5154, 5387]],\n",
    " 'small_T': [[4827, 5387]]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9617790e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mcvList = []\n",
    "for gene in mcvDic:\n",
    "    for each in mcvDic[gene]:\n",
    "        mcvList.append([gene,each[0],each[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "62c2bda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mcvDf = pd.DataFrame(mcvList, columns = ['gene','start','end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "33050895",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VP2</td>\n",
       "      <td>465</td>\n",
       "      <td>1190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VP1</td>\n",
       "      <td>1156</td>\n",
       "      <td>2427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>large_T</td>\n",
       "      <td>2503</td>\n",
       "      <td>4722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large_T</td>\n",
       "      <td>5154</td>\n",
       "      <td>5387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>small_T</td>\n",
       "      <td>4827</td>\n",
       "      <td>5387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  start   end\n",
       "0      VP2    465  1190\n",
       "1      VP1   1156  2427\n",
       "2  large_T   2503  4722\n",
       "3  large_T   5154  5387\n",
       "4  small_T   4827  5387"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcvDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "79dc9980",
   "metadata": {},
   "outputs": [],
   "source": [
    "#MCV gene directions\n",
    "direction = ['+','+','-','-','-']\n",
    "mcvDf['direction'] = ['+','+','-','-','-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "cac26511",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "newStart = []\n",
    "newEnd = []\n",
    "for each in direction:\n",
    "    \n",
    "    if each == '-':\n",
    "        newStart.append(mcvList[i][2])\n",
    "        newEnd.append(mcvList[i][1])\n",
    "    else:\n",
    "        newStart.append(mcvList[i][1])\n",
    "        newEnd.append(mcvList[i][2])\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6b0a183b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mcvDf['newStart'] = newStart\n",
    "mcvDf['newEnd'] = newEnd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e195a273",
   "metadata": {},
   "outputs": [],
   "source": [
    "mcvDf = mcvDf.drop(['start','end'],axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "21d0e501",
   "metadata": {},
   "outputs": [],
   "source": [
    "mcvDf.to_csv('mcvDf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "61440f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "siteDic = {}\n",
    "sampleList = ['83108','83092']\n",
    "sampleName = ['HNSpore4055T','HNSpore4003T']\n",
    "i = 0\n",
    "for sample in sampleList:\n",
    "    path = f'{workdir}/Sample_{sample}/call_fusion_virus'\n",
    "    res = f'{path}/filteredSelectedContig.pickle'\n",
    "    with open(res,'rb') as res:\n",
    "        resDic = pickle.load(res)\n",
    "        if resDic != {}:\n",
    "            siteDic[sampleName[i]] = resDic\n",
    "            \n",
    "    i += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "66295faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneDic = {}\n",
    "gtfFileGenome = 'Homo_sapiens.GRCh38.105.transcript.gtf'\n",
    "with open(gtfFileGenome) as genome:\n",
    "    rows = genome.read().rstrip().split('\\n')\n",
    "    for eachRow in rows:\n",
    "        hasGeneName = False\n",
    "        for each in eachRow.split('\\t')[8].split(';'):\n",
    "            if each.split('\"')[0] == ' gene_name ':\n",
    "                geneName = each.split('\"')[1]\n",
    "                hasGeneName = True\n",
    "            if each.split('\"')[0] == ' transcript_id ':\n",
    "                trID = each.split('\"')[1]\n",
    "        if not hasGeneName:\n",
    "            geneName = trID\n",
    "        chro = eachRow.split('\\t')[0]\n",
    "        start = eachRow.split('\\t')[3]\n",
    "        end = eachRow.split('\\t')[4]\n",
    "        strand = eachRow.split('\\t')[6]\n",
    "        if trID not in geneDic:\n",
    "            geneDic[trID] = [geneName,chro,start,end,strand]\n",
    "        else:\n",
    "            print(eachRow)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1c869f17",
   "metadata": {},
   "outputs": [],
   "source": [
    "genePrimaryList = []\n",
    "geneSiteList = []\n",
    "count = 0\n",
    "for sample in siteDic:\n",
    "    for ins in siteDic[sample]:\n",
    "        chro = ins.split('.')[0]\n",
    "        site = ins.split('.')[1]\n",
    "        geneNameDic = {}\n",
    "        flag = False\n",
    "        for gene in geneDic:\n",
    "            if geneDic[gene][1] == chro and int(geneDic[gene][2]) <= int(site) and int(geneDic[gene][3]) > int(site):\n",
    "                start = geneDic[gene][2]\n",
    "                end = geneDic[gene][3]\n",
    "                dirct = geneDic[gene][4]\n",
    "                geneNameDic[geneDic[gene][0]] = [start,end,dirct]\n",
    "                flag = True\n",
    "        genePrimaryList += geneNameDic.keys()\n",
    "        for gene in geneNameDic:\n",
    "            chro = ins.split('.')[0]\n",
    "            site = ins.split('.')[1]\n",
    "            geneSiteList.append([sample,gene,chro,site] + geneNameDic[gene])\n",
    "        if flag == True:\n",
    "            count += 1\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "97c35b92",
   "metadata": {},
   "outputs": [],
   "source": [
    "countDic = dict(Counter(genePrimaryList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4be688fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genePlotDfList = []\n",
    "for each in geneSiteList:\n",
    "    genePlotDfList.append(each)\n",
    "genePlotDf = pd.DataFrame(genePlotDfList,columns= ['sample','gene','chr','ins','start','end','direction'])\n",
    "genePlotDf.to_csv('geneTrack_MCV.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "03760626",
   "metadata": {},
   "outputs": [],
   "source": [
    "script = 'extractContig.sh'\n",
    "with open(script,'w') as scriptfile:\n",
    "    for each in genePlotDfList:\n",
    "        chro = each[2]\n",
    "        ins = each[3]\n",
    "        sample = each[0]\n",
    "        gene = each[1]\n",
    "        path = work_dir #working dir of searcHPV\n",
    "        contigfile = f'{path}/call_fusion_virus/ContigsSequence.fa'\n",
    "        resfile = f'{path}/{gene}.fa'\n",
    "        scriptfile.write(f'grep -A2 {chro}.{ins} {contigfile} >> {resfile};\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "26d3bae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "script = 'alignContig.sh'\n",
    "with open(script,'w') as scriptfile:\n",
    "    \n",
    "    scriptfile.write('''module load Bioinformatics;\n",
    "module load bwa;\n",
    "module load samtools;\n",
    "''')\n",
    "    for each in genePlotDfList:\n",
    "        gene = each[1]\n",
    "        sample = each[0]\n",
    "        path = 'workdir'\n",
    "        resfile = f'{path}/{gene}.fa'\n",
    "        ref = 'hg_mcv.fa'\n",
    "        scriptfile.write(f'''bwa mem -R '@RG\\\\tID:hpv\\\\tSM:mcv\\\\tLB:mcv\\\\tPL:ILLUMINA' -M -t 8 {ref} {resfile} > {path}/{gene}.sam;\n",
    "    samtools view -bhS {path}/{gene}.sam > {path}/{gene}.bam;\n",
    "    samtools sort -@ 8 -o {path}/{gene}.sort.bam {path}/{gene}.bam;\n",
    "    samtools index {path}/{gene}.sort.bam;''')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "33585edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#exon dictionary\n",
    "file = \"=../data/Homo_sapiens.GRCh38.105.exon.gtf\"\n",
    "exonDic = {}\n",
    "with open(file) as inputFile:\n",
    "    inputFile.readline()\n",
    "    rows = inputFile.read().split('\\n')\n",
    "    for row in rows:\n",
    "        if row != '':\n",
    "            hasExonNum = False\n",
    "            hasGeneName = False\n",
    "            for each in row.split('\\t')[8].split(';'):\n",
    "                if each.split('\"')[0] == ' exon_number ':\n",
    "                    exonNum = each.split('\"')[1]\n",
    "                    hasExonNum = True\n",
    "                if each.split('\"')[0] == ' gene_name ':\n",
    "                    geneName = each.split('\"')[1]\n",
    "                    hasGeneName = True\n",
    "                if each.split('\"')[0] == ' transcript_id ':\n",
    "                    trID = each.split('\"')[1]\n",
    "                if each.split('\"')[0] == ' exon_id ':\n",
    "                    exonID = each.split('\"')[1]\n",
    "            if not hasGeneName:\n",
    "                geneName = trID\n",
    "            if hasExonNum:\n",
    "                exonStart = row.split('\\t')[3]\n",
    "                exonEnd = row.split('\\t')[4]\n",
    "                for each in genePlotDfList:\n",
    "                    if each[1] == geneName:\n",
    "                        exonDic[exonID] = [geneName,exonStart,exonEnd,exonNum,trID]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e65f912",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#open Sample_2655-CB-90 HEMGN contigs alignment\n",
    "annotation = []\n",
    "inserts_annotation = []\n",
    "for eachline in genePlotDfList:\n",
    "    sample = eachline[0]\n",
    "    gene = eachline[1]\n",
    "    # if sample == 'Sample_2655-CB-89' and gene == 'KLF13':\n",
    "    dirct = eachline[6]\n",
    "    ins = eachline[3]\n",
    "    path = 'workdir'\n",
    "    bam = f'{path}/{gene}.sort.bam'\n",
    "    chro = eachline[2]\n",
    "    genestart = eachline[4]\n",
    "    geneend = eachline[5]\n",
    "    genedir = eachline[6]\n",
    "    samfile = pysam.AlignmentFile(bam,'rb')\n",
    "    inserts = []\n",
    "\n",
    "    for read in samfile.fetch(chro,int(genestart),int(geneend)):\n",
    "        start  = read.pos\n",
    "        length = read.query_alignment_length\n",
    "        fullLength = read.infer_read_length()\n",
    "        name = read.qname\n",
    "        contig = name.split('.')[2]\n",
    "        contigins = name.split('.')[1]\n",
    "        #print(sample,chro,ins)\n",
    "        for eachContig in siteDic[sample][chro + '.' + contigins]:\n",
    "            if eachContig[0] == contig:\n",
    "                if eachContig[-1] != 'lowConfidence':\n",
    "                    mcvSite = eachContig[-1]\n",
    "                else:\n",
    "                    mcvSite = eachContig[-2]\n",
    "\n",
    "        mcvGeneBreak = ''\n",
    "        for eachMCVgene in mcvDic:\n",
    "            for eachSlot in mcvDic[eachMCVgene]:\n",
    "                if int(mcvSite) >= int(eachSlot[0]) and int(mcvSite) <= int(eachSlot[1]):\n",
    "                    mcvGeneBreak = mcvGeneBreak + eachMCVgene + ';'\n",
    "        if read.is_reverse:\n",
    "            strand = '-'\n",
    "        else:\n",
    "            strand = '+'\n",
    "        #add flag\n",
    "        flag = 'intron'\n",
    "        exonNumList = ''\n",
    "        for exon in exonDic:\n",
    "            exonStart = int(exonDic[exon][1])\n",
    "            exonEnd = int(exonDic[exon][2])\n",
    "            exonNum = int(exonDic[exon][3])\n",
    "            \n",
    "            if start <=  int(exonEnd) and start + length >= exonStart:\n",
    "                flag = 'exon'\n",
    "                exonNumList = exonNumList + exon + ';'\n",
    "                \n",
    "        \n",
    "               \n",
    "                    \n",
    "        #resolve contig hum components\n",
    "        cigar = read.cigar\n",
    "        infoList = [sample,gene,name,contig,'hum',flag,genedir,fullLength,mcvSite,'',mcvGeneBreak,contigins,genestart,geneend,length,exonNumList,strand]\n",
    "    \n",
    "        \n",
    "        if cigar[0][0] == 0:\n",
    "            #matched part, if strand of contig is different from gene direction, reverse compliment the cigar, switch the start and end\n",
    "            if infoList[6] == '+':\n",
    "                pointStart = 0\n",
    "                pointEnd = cigar[0][1]\n",
    "            else:\n",
    "                pointStart = fullLength\n",
    "                pointEnd = fullLength - cigar[0][1]\n",
    "                \n",
    "        else:\n",
    "            pointStart = cigar[1][1]\n",
    "            if cigar[-1][0] == 4 or cigar[-1][0] == 5:\n",
    "                pointEnd = fullLength - cigar[-1][1]\n",
    "            else:\n",
    "                pointEnd = fullLength\n",
    "            if infoList[6] == '-':\n",
    "                pointStart = fullLength - cigar[0][1]\n",
    "                pointEnd = fullLength - pointEnd\n",
    "        inserts.append([pointStart,pointEnd] + infoList)\n",
    "        \n",
    "        \n",
    "        if read.has_tag('SA'):\n",
    "            saTag = read.get_tag('SA')\n",
    "        else:\n",
    "            print(read)\n",
    "        #print(saTag)\n",
    "        aligns = saTag.split(';')\n",
    "        for align in aligns:\n",
    "            if align != '':\n",
    "                supchro = align.split(',')[0]\n",
    "                if supchro == 'NC_010277.2':\n",
    "                    suppos = align.split(',')[1]\n",
    "                    supstrand = align.split(',')[2]\n",
    "                    supcigar = align.split(',')[3]\n",
    "                    slots = re.split(r'[A-Z]',supcigar)[:-1]\n",
    "                    slotCigars = re.split(r'\\d+',supcigar)[1:]\n",
    "                    #print(align,slots,slotCigars)\n",
    "                    i = 0\n",
    "                    #add flag\n",
    "                    if slotCigars[0] == 'M':\n",
    "                        match = int(slots[0])\n",
    "                        supEnd = mcvSite - match\n",
    "                    else:\n",
    "                        match = int(slots[1])\n",
    "                        supEnd = mcvSite + match\n",
    "                    newFlag = []\n",
    "                    hasEach1 = False\n",
    "                    for each in mcvDic:\n",
    "                        for eachone in mcvDic[each]:\n",
    "                            if int(suppos) >= int(eachone[0]) and int(suppos) <= int(eachone[1]):\n",
    "                                each1 = each\n",
    "                                newFlag.append(each1)\n",
    "                                hasEach1 = True\n",
    "                            if int(suppos) + match >= int(eachone[0]) and int(suppos) + match <= int(eachone[1]):\n",
    "                                each2 = each\n",
    "                                if each2 != each1 and hasEach1:\n",
    "                                    newFlag.append(each2)\n",
    "                                elif not hasEach1:\n",
    "                                    newFlag.append(each2)\n",
    "\n",
    "                    if newFlag == []:\n",
    "                        newFlag = ''\n",
    "                    else:\n",
    "                        newFlag = ';'.join(newFlag)\n",
    "                    #print(newFlag)\n",
    "                    infoList = [sample,gene,name,contig,'mcpv',newFlag,genedir,fullLength,mcvSite,supEnd,mcvGeneBreak,contigins,genestart,geneend,length,'',supstrand]\n",
    "                    if slotCigars[0] == 'M':\n",
    "                        if supstrand == strand: #read converted to consistent cigar to original human alignment\n",
    "                            if infoList[6] == '+': #read on hum refence was not converted\n",
    "                                pointStart = 0\n",
    "                                pointEnd = int(slots[0])\n",
    "                            else: #read on hum refence was not converted\n",
    "                                pointStart = fullLength\n",
    "                                pointEnd = fullLength - int(slots[0])\n",
    "                        else: #read have inverted cigar to original human alignment\n",
    "                            if infoList[6] == '+': #read on hum refence was not converted, converted to human alignments\n",
    "                                pointStart = fullLength\n",
    "                                pointEnd = fullLength - int(slots[0])\n",
    "                            else:#read on hum refence was not converted, don't converted\n",
    "                                pointStart = 0\n",
    "                                pointEnd = int(slots[0])\n",
    "                                \n",
    "                        inserts.append([pointStart,pointEnd] + infoList)\n",
    "                        #print(supcigar,supstrand,strand)\n",
    "                        #print([pointStart,pointEnd] + infoList)\n",
    "                    else:\n",
    "\n",
    "                        if supstrand == strand: #read converted to consistent cigar to original human alignment\n",
    "                            pointStart = int(slots[0])\n",
    "                            if slotCigars[-1] == 'S' or slotCigars[-1] == 'H':\n",
    "                                pointEnd = fullLength - int(slots[-1])\n",
    "                            else:\n",
    "                                pointEnd = fullLength\n",
    "                            if infoList[6] == '-': #read on hum refence was converted\n",
    "                                pointStart = fullLength - pointStart\n",
    "                                pointEnd = fullLength - pointEnd\n",
    "                               \n",
    "                        else:\n",
    "\n",
    "                            if slotCigars[-1] == 'S' or slotCigars[-1] == 'H':\n",
    "                                pointStart = int(slots[-1])\n",
    "                            else:\n",
    "                                pointStart = 0\n",
    "                            pointEnd = fullLength - int(slots[0])\n",
    "                            #base on direction change order of start and end\n",
    "                            #print(infoList[6])\n",
    "                            #print(pointStart)\n",
    "                            if infoList[6] == '-': #read on hum refence was not converted, converted to human alignments\n",
    "                                pointStart = fullLength - pointStart\n",
    "                                pointEnd = fullLength - pointEnd\n",
    "                            inserts.append([pointEnd,pointStart] + infoList)\n",
    "                            \n",
    "        \n",
    "\n",
    "        annotation.append([sample,gene,sample,contig,chro,start,length,strand,contigins])\n",
    "        #print(inserts)\n",
    "    #print(suppos,int(suppos) + match,inserts)\n",
    "    inserts_annotation += inserts\n",
    "    #pd.DataFrame(inserts,columns = ['name','contig','start','end','type','feature','geneDirection','fullLength']).to_csv(outputFile2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "fa7caa91",
   "metadata": {},
   "outputs": [],
   "source": [
    "outputFile1 = 'annotation.csv'\n",
    "outputFile2 = 'inserts.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "a52c9e36",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_df = pd.DataFrame(annotation,columns = ['sampleID','gene','sampleName','contig','chr','start','length','strand','ins'])\n",
    "annotation_df.to_csv(outputFile1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "9e4687cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "inserts_df = pd.DataFrame(inserts_annotation,columns = ['start','end','sample','gene','name','contig','type','feature','geneDirection','fullLength','hpvSite','hpvEnd','hpvGene','ins','geneStart','geneEnd','length','exonNum','strand'] )\n",
    "inserts_df.to_csv(outputFile2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
